src/vcpkg/base/files.cpp

```diff
@@ -721,8 +679,65 @@ namespace
     };
 
     PosixDType get_d_type(const struct dirent*) noexcept { return PosixDType::Unknown; }
-#endif
+#endif // ^^^ !_DIRENT_HAVE_D_TYPE
 
+    void vcpkg_remove_all(const Path& base, std::error_code& ec, Path& failure_point)
+    {
+        {
+            ReadDirOp op{base, ec};
+            if (!ec)
+            {
+                // it was a directory, so delete everything inside
+                for (;;)
+                {
+                    errno = 0;
+                    auto entry = op.read();
+                    if (!entry)
+                    {
+                        if (errno != 0)
+                        {
+                            ec.assign(errno, std::generic_category());
+                            failure_point = base;
+                            return;
+                        }
+
+                        // no more entries left, fall down to unlink below
+                        break;
+                    }
+
+                    // delete base / entry.d_name, recursively
+                    if (is_dot_or_dot_dot(entry->d_name))
+                    {
+                        continue;
+                    }
+
+                    vcpkg_remove_all(base / entry->d_name, ec, failure_point);
+                    if (ec)
+                    {
+                        // removing a contained entity failed; the recursive call will set failure_point
+                        return;
+                    }
+                }
+            }
+            else if (ec == std::errc::not_a_directory)
+            {
+                // was not a directory, fall down to the unlink below
+                ec.clear();
+            }
+            else
+            {
+                // some other IO error occurred trying to open the directory
+                failure_point = base;
+                return;
+            }
+        } // close op
+
+        if (::unlink(base.c_str()) != 0)
```

Comment:
I don't think `::unlink()` supports deleting directories? According to the Linux manpage:

> `EISDIR`: `pathname` refers to a directory.  (This is the non-POSIX value returned by Linux since 2.1.132.

is one of the errnos that can be returned by `::unlink`, which implies to me that `::unlink` cannot remove directories.

Additionally, when testing it on my own personal Mac, the following C:

```c
int unlink(const char*);
int printf(const char*, ...);
extern int errno;

int main(int argc, char** argv) {
	if (argc != 2)
	{
		printf("%s requires one argument\n", argv[0]);
		return 1;
	}
	int res = unlink(argv[1]);
	printf("unlink(%s) returned %d, errno = %d\n", argv[1], res, errno);
	return 0;
}
```

results in the following:

```sh
$ mkdir testdir
$ touch testfile
$ ./a.out      
$ ./a.out testdir
unlink(testdir) returned -1, errno = 1 # note: EPERM
$ touch testfile 
$ ./a.out testfile
unlink(testfile) returned 0, errno = 22
```